examples/td3/td3.py

```diff
@@ -2,302 +2,126 @@
+        buffer_size=cfg.replay_buffer.size,
         device=device,
     )
 
-    # Optimizers
-    critic_params = list(loss_module.qvalue_network_params.flatten_keys().values())
-    actor_params = list(loss_module.actor_network_params.flatten_keys().values())
-
-    optimizer_actor = optim.Adam(actor_params, lr=cfg.lr, weight_decay=cfg.weight_decay)
-    optimizer_critic = optim.Adam(
-        critic_params, lr=cfg.lr, weight_decay=cfg.weight_decay
-    )
+    # Make Optimizers
+    optimizer_actor, optimizer_critic = make_optimizer(cfg, loss_module)
 
     rewards = []
     rewards_eval = []
 
     # Main loop
     collected_frames = 0
-    pbar = tqdm.tqdm(total=cfg.total_frames)
+    pbar = tqdm.tqdm(total=cfg.collector.total_frames)
     r0 = None
     q_loss = None
 
+    init_random_frames = cfg.collector.init_random_frames
+    num_updates = int(
+        cfg.collector.env_per_collector
+        * cfg.collector.frames_per_batch
+        * cfg.optimization.utd_ratio
+    )
+    delayed_updates = cfg.optimization.policy_update_delay
+    prb = cfg.replay_buffer.prb
+    env_per_collector = cfg.collector.env_per_collector
+    eval_rollout_steps = cfg.collector.max_frames_per_traj // cfg.env.frame_skip
+    eval_iter = cfg.logger.eval_iter
+    frames_per_batch, frame_skip = cfg.collector.frames_per_batch, cfg.env.frame_skip
+
     for i, tensordict in enumerate(collector):
+        exploration_policy.step(tensordict.numel())
         # update weights of the inference policy
         collector.update_policy_weights_()
 
         if r0 is None:
             r0 = tensordict["next", "reward"].sum(-1).mean().item()
         pbar.update(tensordict.numel())
 
-        # extend the replay buffer with the new data
-        tensordict = tensordict.view(-1)
+        tensordict = tensordict.reshape(-1)
         current_frames = tensordict.numel()
         replay_buffer.extend(tensordict.cpu())
         collected_frames += current_frames
 
         # optimization steps
-        if collected_frames >= cfg.init_random_frames:
+        if collected_frames >= init_random_frames:
             (
                 actor_losses,
                 q_losses,
             ) = ([], [])
-            for i in range(
-                int(cfg.env_per_collector * cfg.frames_per_batch * cfg.utd_ratio)
-            ):
+            for j in range(num_updates):
                 # sample from replay buffer
-                sampled_tensordict = replay_buffer.sample(cfg.batch_size).clone()
+                sampled_tensordict = replay_buffer.sample().clone()
 
                 loss_td = loss_module(sampled_tensordict)
 
                 actor_loss = loss_td["loss_actor"]
                 q_loss = loss_td["loss_qvalue"]
 
                 optimizer_critic.zero_grad()
-                update_actor = i % cfg.policy_update_delay == 0
+                update_actor = j % delayed_updates == 0
                 q_loss.backward(retain_graph=update_actor)
```

Comment:
Do we need the retain_graphs?
In theory we should not, unless there's a bug. They should be separate graphs.